### PR TITLE
Update datatables.js from 1.10.5 to 1.10.22 and deprecate `renderDataTable()`/`dataTableOutput()`

### LINKED ISSUE #3627: Security: DataTables.net prototype pollution
### System details

Browser Version: 

Output of `sessionInfo()`:

```
# sessionInfo() output goes here
```

### Example application *or* steps to reproduce the problem

 

```R
# Minimal, self-contained example app code goes here
```

### Describe the problem in detail
EXPLANATION
The datatables.net package is vulnerable to Prototype Pollution. The setData function in jquery.dataTables.js fails to protect prototype attributes when objects are created during the application's execution. A remote attacker can exploit this to modify the behavior of object prototypes which, depending on their use in the application, may result in a Denial of Service (DoS), Remote Code Execution (RCE), or other unexpected execution flow.

DETECTION
The application is vulnerable by using this component.

ROOT CAUSE
shiny_1.7.1.tar.gzshiny/inst/www/shared/datatables/js/jquery.dataTables.min.js( ,1.10.13)
shiny_1.7.1.tar.gzshiny/inst/www/shared/datatables/js/jquery.dataTables.min.js(,)

### LINKED ISSUE #3986: Update bundled DataTable.js; deprecate `dataTableOutput()`
The bundled datatables.js is getting flagged by various customer CVE tools. I think it's worth updating the currently bundled script (it looks like there's a bundled script to do so) and move `dataTableOutput()` towards more aggressive deprecation so the JS can be removed all together.

(Oh the script exists but says that DataTables won't get updated again 😬)

### PR DESCRIPTION
**Pull request by cpsievert**
Closes #3627 
Closes #3986
Related https://github.com/rstudio/DT/pull/1130, https://github.com/rstudio/DT/pull/1128, and https://github.com/rstudio/DT/pull/1132

This PR updates datatables.js from v1.10.5 to v1.10.22 to address the security issue discussed in #3627. 

It also makes it so `renderDataTable()`/`dataTableOutput()` throw a deprecation message to use their `{DT}` equivalents. Also, if a suitable (upcoming) version of `{DT}` is available, they automatically dispatch to their `{DT}` equivalents (`renderDT`/`DTOutput()`). In this case, the user will also see a message about setting `options(shiny.legacy.datatable = TRUE)` if they'd still like the old implementation.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## Breaking changes
 
+* `renderDataTable()`/`dataTableOutput()` have been officially deprecated and will automatically dispatch to their `{DT}` equivalents (`renderDT()`/`DTOutput()`) when a recent version of `{DT}` is available. If this happens to degrade an existing app, set `options(shiny.legacy.datatables = TRUE)` to get the old implementation. (#3998)
```
Comment (cpsievert) [inline NEWS.md:5]:
```suggestion
* `renderDataTable()`/`dataTableOutput()` have been officially deprecated and will automatically dispatch to their `{DT}` equivalents (`renderDT()`/`DTOutput()`) when a recent version of `{DT}` is available. If this happens to degrade an existing app, set `options(shiny.legacy.datatable = TRUE)` to get the old implementation. (#3998)
```

File: inst/www/shared/datatables/js/jquery.dataTables.min.js [hunk 2]
```diff
@@ -1,158 +1,184 @@
...
+typeof a||-1===a.indexOf(".")&&-1===a.indexOf("[")&&-1===a.indexOf("("))return function(d,f){return d[a]};var c=function(d,f,e){if(""!==e){var g=ab(e);for(var h=0,l=g.length;h<l;h++){e=g[h].match(ua);var n=g[h].match(oa);if(e){g[h]=g[h].replace(ua,"");""!==g[h]&&(d=d[g[h]]);n=[];g.splice(0,h+1);g=g.join(".");if(Array.isArray(d))for(h=0,l=d.length;h<l;h++)n.push(c(d[h],f,g));d=e[0].substring(1,e[0].length-1);d=""===d?n:n.join(d);break}else if(n){g[h]=g[h].replace(oa,"");d=d[g[h]]();continue}if(null===
+d||d[g[h]]===q)return q;d=d[g[h]]}}return d};return function(d,f){return c(d,f,a)}}function da(a){if(k.isPlainObject(a))return da(a._);if(null===a)return function(){};if("function"===typeof a)return function(c,d,f){a(c,"set",d,f)};if("string"!==typeof a||-1===a.indexOf(".")&&-1===a.indexOf("[")&&-1===a.indexOf("("))return function(c,d){c[a]=d};var b=function(c,d,f){f=ab(f);var e=f[f.length-1];for(var g,h,l=0,n=f.length-1;l<n;l++){if("__proto__"===f[l])throw Error("Cannot set prototype values");g=
```
Comment (cpsievert) [inline inst/www/shared/datatables/js/jquery.dataTables.min.js:36]: For reference this line contains the prototype pollution safeguard https://github.com/DataTables/Dist-DataTables/commit/e2e19eac7e5a6f140d7eefca5c7deba165b357eb

**Review by jcheng5 — APPROVED** [review]

File: NEWS.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 ## Breaking changes
 
+* `renderDataTable()`/`dataTableOutput()` have been officially deprecated and will automatically dispatch to their `{DT}` equivalents (`renderDT()`/`DTOutput()`) when a recent version of `{DT}` is available. If this happens to degrade an existing app, set `options(shiny.legacy.datatable = TRUE)` to get the old implementation. (#3998)
```
Comment (jcheng5) [inline NEWS.md:5]: Perhaps make it clearer in the first sentence that people should switch to the DT equivalents?

Comment (cpsievert) [inline NEWS.md:5]:
```suggestion
* `renderDataTable()`/`dataTableOutput()` are officially deprecated in favor of [their `{DT}` equivalents](https://rstudio.github.io/DT/shiny.html), `renderDT()`/`DTOutput()`. In most cases, simply changing `renderDataTable()` to `DT::renderDT()` and `dataTableOutput()` to `DT::DTOutput()` is sufficient for upgrading. Also, to promote migration, when a recent version of `{DT}` is available, `renderDataTable()`/`dataTableOutput()` now automatically use their `{DT}` equivalent (and provide a message that they are doing so). If this happens to degrade an existing app, set `options(shiny.legacy.datatable = TRUE)` to get the old (i.e., non-`{DT}`) implementation. (#3998)
```

**Pull request merged**